src/checker/ruleset_location.go

```diff
@@ -0,0 +1,179 @@
+	err := os.MkdirAll(r.getRuleSetPath(), 0700)
+	if errors.Is(err, fs.ErrExist) {
+		return nil
+	}
+	return err
+}
+
+func (r *RuleSetLocation) getFullUrl() string {
+	return "https://" + r.simpleUrl
+}
+
+func (r *RuleSetLocation) getRuleSetPath() string {
+	dir, err := getRuleSetsDir()
+	if err != nil {
+		return ""
+	}
+	return dir + "/" + r.simpleUrl
+}
+
+func (r *RuleSetLocation) isRuleSetDownloaded() bool {
+	_, err := os.ReadDir(r.getRuleSetPath())
+	return !errors.Is(err, fs.ErrNotExist)
+}
+
+func (rulesetLocation *RuleSetLocation) createLockfile() ([]string, error) {
+	configPath := rulesetLocation.getRuleSetPath() + "/" + CONFIG_FILENAME
+	configFileContents, error := os.ReadFile(configPath)
+	if error != nil {
+		return []string{}, error
+	}
+
+	type DependencyTreeLayout struct {
+		Dependencies []string
+	}
+	var basicDependencyTree DependencyTreeLayout
+
+	error = yaml.Unmarshal(configFileContents, &basicDependencyTree)
+	if error != nil {
+		return []string{}, error
+	}
+
+	var outputDependencyTree DependencyTreeLayout
+	outputDependencyTree.Dependencies = make([]string, len(basicDependencyTree.Dependencies))
+	copy(outputDependencyTree.Dependencies, basicDependencyTree.Dependencies)
+
+	for _, dependencyName := range basicDependencyTree.Dependencies {
+		dependencyLocation := RuleSetLocation{}
+		dependencyLocation.new(strings.Split(dependencyName, "@")[0])
+		if !dependencyLocation.isRuleSetDownloaded() {
+			error = FetchRuleSet(&dependencyLocation)
+			if error != nil {
+				return []string{}, error
+			}
+		}
+		lockfilePath := dependencyLocation.getRuleSetPath() + "/" + LOCK_FILENAME
+		if api.FileExists(lockfilePath, false) {
+			temporaryDependencyTree := DependencyTreeLayout{}
+			configFileContents, error = os.ReadFile(dependencyLocation.getRuleSetPath() + "/" + CONFIG_FILENAME)
+			if error != nil {
+				return []string{}, error
+			}
+
+			error = yaml.Unmarshal(configFileContents, &temporaryDependencyTree)
+
+			if error != nil {
+				return []string{}, error
+			}
+			for _, nestedDependencyName := range temporaryDependencyTree.Dependencies {
+				if !slices.Contains(outputDependencyTree.Dependencies, nestedDependencyName) {
+					 outputDependencyTree.Dependencies = append(outputDependencyTree.Dependencies, nestedDependencyName)
+				}
+			}
+		} else {
+			recursiveDependencyTree, error := dependencyLocation.createLockfile()
+			if error == nil {
+				outputDependencyTree.Dependencies = append(outputDependencyTree.Dependencies, recursiveDependencyTree...)
+			}
+		}
+	}
+
+	lockfilePath := rulesetLocation.getRuleSetPath() + "/" + LOCK_FILENAME
+	lockfile, error := os.Create(lockfilePath)
+	
+	if error != nil {
+		return []string{}, error
+	}
+	defer lockfile.Close()
+
+	yamlOutput, error := yaml.Marshal(outputDependencyTree)
+	if error != nil {
+		return []string{}, error
+	}
+
+	lockfile.Write(yamlOutput)
+
+	return outputDependencyTree.Dependencies, nil
+}
+
+func anyToError(val any) error {
```

Comment:
Apart from the fact that it seems weird to create new utils file only for this function, I do not think `anyToError` should placed in this file especially that it isn't even used in `ruleset_location.go`